[Model] Add PaddleOCR-VL Model Support

zhang-prog opened the pull request:
# What does this PR do?
This PR adds PaddleOCR-VL model to Hugging Face Transformers from PaddleOCR.

Relevant Links:

[PaddleOCR](https://www.paddleocr.ai/latest/index.html)
https://huggingface.co/PaddlePaddle/PaddleOCR-VL

## Usage

### Use a pipeline

```python
from transformers import pipeline

pipe = pipeline("image-text-to-text", model="PaddlePaddle/PaddleOCR-VL", dtype="bfloat16")
messages = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/ocr_demo2.jpg"},
            {"type": "text", "text": "OCR:"},
        ]
    }
]
result = pipe(text=messages)
print(result)
```

### Load model directly

```python
from transformers import AutoProcessor, AutoModelForImageTextToText

processor = AutoProcessor.from_pretrained("PaddlePaddle/PaddleOCR-VL")
model = AutoModelForImageTextToText.from_pretrained("PaddlePaddle/PaddleOCR-VL", dtype="bfloat16")
messages = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/ocr_demo2.jpg"},
            {"type": "text", "text": "OCR:"},
        ]
    }
]
inputs = processor.apply_chat_template(
	messages,
	add_generation_prompt=True,
	tokenize=True,
	return_dict=True,
	return_tensors="pt",
).to(model.device)

outputs = model.generate(**inputs, max_new_tokens=100)
result = processor.decode(outputs[0][inputs["input_ids"].shape[-1]:-1])
print(result)

```

zhang-prog committed:
init

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+    if width < factor:
+        height = round((height * factor) / width)
+        width = factor
+
```
same as Qwen but with support for H/W smaller than a factor. I think we made qwen-VL support small images as well, so prob directly importing will give expected result?

zucchini-nlp replied:
You mean the default values of min-max pixels? We're passing the correct value from within image processor when calling `smart_resize`, so it shouldn't be a problem, no?

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+    tokenizer_class = "AutoTokenizer"
+
+    def __init__(self, image_processor=None, tokenizer=None, chat_template=None, **kwargs):
+        self.image_token = "<|IMAGE_PLACEHOLDER|>" if not hasattr(tokenizer, "image_token") else tokenizer.image_token
```
can you add the token in tokenizer so we can assume t's always available?
https://huggingface.co/docs/transformers/en/main_classes/tokenizer#multimodal-tokenizer

zhang-prog replied:
I tried adding `image_token` to the `special_tokens_map.json`, but found it had no effect. What should I do? Do you have an example?

zucchini-nlp replied:
can you try with the above link? It should work unless smth got broken in the latest refactor

zhang-prog replied:
Sure. We add `image_token` to `tokenizer_config.json` to solve this problem

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+        if images is not None:
+            image_inputs = self.image_processor(images=images, return_tensors="pt")
+            image_inputs["pixel_values"] = image_inputs["pixel_values"]
```
dummy no-op line?

zhang-prog replied:
Removed

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        temporal_patch_size=2,
+        tokens_per_second=2,
+        **kwargs,
+    ):
```
we can use modular and copy from any vision config since most attributes are common and re-define model-specific ones. If it's not possible we need a docstring

zucchini-nlp replied:
comment not addressed

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+    model_type = "paddleocr_vl"
+    keys_to_ignore_at_inference = ["past_key_values"]
+    sub_configs = {"vision_config": PaddleOCRVisionConfig}
```
has be be nested with text and vision configs, similar to qwen and other VLMs. Half-nested configs can have side effects as we assume a complete nested or complete flat config format throughout the repo

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        max_sequence_length=None,
+        tie_word_embeddings=False,
+        vision_config=None,
+        rope_scaling=None,
```
it was renamed to `rope_parameters` in v5. Now we don't have a separate `rope_theta` and store all RoPE-related args in "rope_parameters" as a dict

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        super().__init__(tie_word_embeddings=tie_word_embeddings, **kwargs)
+
+
+class Projector(nn.Module):
```
let's call it `PaddlePaddleOCRProjector` and include model name explicitly. Same for other modules if they don't contain model's name

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+    def forward(self, seqlen: int) -> torch.Tensor:
+        seq = torch.arange(seqlen, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
+        freqs = torch.outer(seq, self.inv_freq)
```
looks like simple RoPE, we can inherit from `LlamaRotaryEmbedding`. Let's use standardized RoPE API unless it uses a new rope type

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+            position_ids = kwargs["position_ids"]
+            if position_ids.dim() == 3 and position_ids.shape[0] > 1:
+                kwargs["position_ids"] = position_ids[0:1]
+
```
we can do it in PreTrainedModel's forward and therefore attention's forward will be 100% identical to qwen2-VL

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+            embeddings = rearrange(embeddings, "(b l) d -> b l d", b=batch_size, l=squence_len)
+
+            if image_grid_thw is not None:
+                flatten_image_grid_thw = self.flatten_list(image_grid_thw)
```
`image_grid_thw` is usually a tensor when returned from processors, and we can call `image_grid_thw.flatten()` here

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+            if image_grid_thw is not None:
+                flatten_image_grid_thw = self.flatten_list(image_grid_thw)
+                assert batch_size == 1
```
can't process more than 1 image at a time? If yes, lets' raise a proper error with a clear message

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+                assert sum([np.prod(x) for x in flatten_image_grid_thw]) == embeddings.shape[1], (
+                    flatten_image_grid_thw,
+                    embeddings.shape,
+                )
```
either not needed to check or we need to raise a proper error with a clear and actionable message for users

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        rope_emb: Optional[tuple[torch.Tensor, torch.Tensor]] = None,
```
`rotary_embeddings` so we have standardized naming :)

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        """
+        device = inputs_embeds.device
+        hidden_states = inputs_embeds
+        attention_mask = attention_mask.to(inputs_embeds.dtype) if attention_mask is not None else None
```
i don't think this will make sense if attention is passed as a boolean 2D array. We can prepare attention with `create_bidirectional_mask` from https://github.com/huggingface/transformers/blob/6408d3b01a4b5446655c222daa0b947bddf7db82/src/transformers/masking_utils.py#L866 instead which is more robust to different attention functions

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+            split_hids.append(sample_hids)
+            split_wids.append(sample_wids)
+        width_position_ids = torch.concat(split_wids, dim=0)
+        height_position_ids = torch.concat(split_hids, dim=0)
```
could be vectorizable, up to you though since image_thw is usually a tiny array

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+    def get_decoder(self):
+        return self.model
+
+    def get_rope_index(
```
position ids are same as qwen2-5-VL, so lets inherit from `Qwen25VLModel` when initializing the base model

zhang-prog replied:
Done

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+                        f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
+                    )
+
+                mask = input_ids == self.config.image_token_id
```
these part needs to be split out with helper fn such as `get_image_features` and `get_placeholder_masks`. You can look at qwen-vl or any VLM for an example

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
+
+            if attention_mask is not None:
+                attention_mask = attention_mask.to(inputs_embeds.device)
```
not needed, it's handled by transformers internally

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+            shift_labels = shift_labels.view(-1)
+            # Enable model parallelism
+            shift_labels = shift_labels.to(shift_logits.device)
+            loss = loss_fct(shift_logits, shift_labels)
```
let's use `self.loss_fn` here

zhang-prog replied:
Done

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+        return model_inputs
+
+    def _get_image_nums_and_video_nums(
```
same as qwen2-5-vl, can be deleted after inheriting from it

zucchini-nlp reviewed (commented):
hey @zhang-prog , thanks for the PR! Great model to have in transformers!

The main thing to fix first is the naming, it should clearly include "PaddlePaddleOCR" and follow the usual pattern depending on the modality. The config format also isn’t right; it needs to be fully nested, with text and vision configs inside. Additionally there are no tests or docs, several files are missing. You can run `transformers add-new-model-like` which would generate a placeholder with the necessary files. I also left some smaller comments here and there. Let me know if you hit any issues

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+        return image_nums, video_nums
+
+    def _expand_inputs_for_generation(
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+            rope_deltas=self.rope_deltas,
+        )
+
+    def prepare_inputs_for_generation(
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+class PaddleOCRVLForConditionalGeneration(Ernie4_5PreTrainedModel, GenerationMixin):
+    def __init__(self, config):
+        super().__init__(config)
+        self.mlp_AR = Projector(config, config.vision_config)
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+
+
+class Projector(nn.Module):
+    def __init__(self, text_config: PaddleOCRVLConfig, vision_config: PaddleOCRVisionConfig):
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+    return h_bar, w_bar
+
+
+class PaddleOCRVLImageProcessor(Qwen2VLImageProcessor):
```
Done

zhang-prog committed:
refactor

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+        temporal_patch_size: int = 1,
+        merge_size: int = 2,
+        **kwargs,
+    ) -> None:
```
same for `init`, we can skip the body and leave as `super().__init__`. Modular will unpack it automatically by copying from qwen

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1933 @@
+        hidden_size=768,
+        intermediate_size=11008,
+        max_position_embeddings=32768,
+        num_hidden_layers=2,
```
comment not addressed

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+            pixel_values (`torch.FloatTensor` of shape `(batch_size, sequence_length, image_channels, patch_size, patch_size)`):
+                The tensors corresponding to the input images.
+            position_ids (`torch.LongTensor` of shape `sequence_length`):
+                The position ids of the image.
```
nit: no position ids in the signature

zhang-prog replied:
Removed

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+
+        assert batch_size == 1, (
+            f"Batch size must be 1, but received {batch_size}. This model only processes one image at a time."
+        )
```
the question about bs remains, does that mean we can't support the model with bs > 1? The model need to support common features such as batched inference before we can merge it

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+        self,
+        inputs_embeds,
+        cu_seqlens,
+        attention_mask: Optional[torch.Tensor] = None,
```
nit:  let's add type hints for all args

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+        mask_unsqueezed = mask.unsqueeze(-1)
+        mask_expanded = mask_unsqueezed.expand_as(inputs_embeds)
+        image_mask = mask_expanded.to(inputs_embeds.device)
+        return image_mask
```
we need to support `inputs_embeds` as well. Can be copied from llava

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+
+        if pixel_values is not None:
+            image_embeds = self.get_image_features(pixel_values, image_grid_thw)
+            image_embeds = self.projector(image_embeds, image_grid_thw)
```
`projector` has to be within `self.get_image_features`, because usually `get_image_features` holds all image processing required for a model

zucchini-nlp reviewed (commented):
@zhang-prog thanks for iterating!

There are a couple major comments which were not addressed and can be a blocker for merging. 
1. The model seems to not support batched inference in current state. We need to enable batching before merging if possible. Should not be hard I think given that the image tower is quite similar to existing models
2. We also need tests to make sure everything actually works and a documentation page. These files are usually auto-prefilled with empty files when you run `transformers add-new-model-like`
3. Let the modular copy automatically when possible. I think there are a few more modules which can be copied from similar models. If you struggle with finding a similar model, you can try out a [modular detector](https://github.com/huggingface/transformers/blob/main/utils/modular_model_detector.py)

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+            rope_deltas=self.rope_deltas,
+        )
+
+        return output if return_dict else output.to_tuple()
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+                The attention_mask used in forward function shape [batch_size X sequence_length] if not None.
+            position_ids (`torch.LongTensor` of shape `sequence_length`):
+                The position ids of the image.
+            image_grid_thw (`torch.LongTensor` of shape `(num_images, 3)`, *optional*):
```
Removed

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+
+    def _preprocess(
+        self,
+        images: ImageInput,
```
no, the shapes of  `patches` and `flatten_patches` are not consistent.

```diff
        patches = patches.reshape(
            grid_t,
            temporal_patch_size,
            channel,
-            grid_h // merge_size,
-            merge_size,
-            patch_size,
-            grid_w // merge_size,
-            merge_size,
-            patch_size,
+            grid_h,
+            self.patch_size,
+            grid_w,
+            self.patch_size,
         )
-        patches = patches.transpose(0, 3, 6, 4, 7, 2, 1, 5, 8)
-        flatten_patches = patches.reshape(
-            grid_t * grid_h * grid_w, channel * temporal_patch_size * patch_size * patch_size
-        )
+        patches = patches.transpose(0, 3, 5, 2, 1, 4, 6)
+        assert self.temporal_patch_size == 1 
+        flatten_patches = patches.reshape(grid_t * grid_h * grid_w, channel, self.patch_size, self.patch_size)
         return flatten_patches, (grid_t, grid_h, grid_w)
```

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+
+class PaddleOCRVLImageProcessorFast(BaseImageProcessorFast):
+    def __init__(
+        self,
```
The `_preprocess` is not exactly the same, the specific details are the same as mentioned in the above comment.

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+    def forward(
+        self,
+        pixel_values,
+        cu_seqlens,
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+        super().__init__()
+        self.text_config = config.text_config
+        self.vision_config = config.vision_config
+        self.merge_kernel_size = (2, 2)
```
Done

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+        min_pixels (`int`, *optional*, defaults to `28 * 28 * 130`):
+            The min pixels of the image to resize the image.
+        max_pixels (`int`, *optional*, defaults to `28 * 28 * 1670`):
+            The max pixels of the image to resize the image.
```
Done

zhang-prog committed:
update

zhang-prog commented:
> @zhang-prog thanks for iterating!
> 
> There are a couple major comments which were not addressed and can be a blocker for merging.
> 
> 1. The model seems to not support batched inference in current state. We need to enable batching before merging if possible. Should not be hard I think given that the image tower is quite similar to existing models
> 2. We also need tests to make sure everything actually works and a documentation page. These files are usually auto-prefilled with empty files when you run `transformers add-new-model-like`
> 3. Let the modular copy automatically when possible. I think there are a few more modules which can be copied from similar models. If you struggle with finding a similar model, you can try out a [modular detector](https://github.com/huggingface/transformers/blob/main/utils/modular_model_detector.py)

@zucchini-nlp In response to your suggestions, we have fixed the issues mentioned in the comments and replied to your overall recommendations:

1. We support bs > 1, like this:

```python
from transformers import AutoProcessor, AutoModelForImageTextToText

processor = AutoProcessor.from_pretrained("PaddlePaddle/PaddleOCR-VL")
model = AutoModelForImageTextToText.from_pretrained("PaddlePaddle/PaddleOCR-VL", dtype="bfloat16").to(DEVICE).eval()
messages1 = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/ocr_demo.jpg"},
            {"type": "text", "text": "OCR:"},
        ]
    }
]
messages2 = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/ocr_demo2.jpg"},
            {"type": "text", "text": "OCR:"},
        ]
    }
]
batch_messages = [messages1, messages2]
inputs = processor.apply_chat_template(
	batch_messages,
	add_generation_prompt=True,
	tokenize=True,
	return_dict=True,
	return_tensors="pt",
    padding=True,
    padding_side='left',
).to(model.device)

generated_ids = model.generate(**inputs, max_new_tokens=1024)
generated_ids_trimmed = [out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)]
result = processor.batch_decode(generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False)
print(result)
```

2. We still have some issues to discuss. I replied to your comment and will generate the final version of the document once it’s completed.

3. We also added the `PaddleOCRVisionConfig` and `PaddleOCRTextConfig` into modular. 

Please review the code again when you have time. Thank you for your efforts. ❤️

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1182 @@
+            self.patch_size,
+        )
+        patches = patches.transpose(0, 3, 5, 2, 1, 4, 6)
+        assert self.temporal_patch_size == 1
```
and here as well, `if temporal_patch_size != 1: raise ValueError(informative message suggesting to not pass any other value but 1)`

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1182 @@
+                or self.rope_deltas is None
+                or (past_key_values is None or past_key_values.get_seq_length() == 0)
+            ):
+                position_ids, rope_deltas = self.get_rope_index(
```
qwen-vl has updated building position ids recently due to expectations from PEFT, needs a rebase and update in PaddleOCR

zucchini-nlp reviewed (commented):
Nice, only a few comments and replied to your questions above

For the docs and the tests, they need to be in `source/docs/en/model_doc` and in `tests` folder. You can take a look at the recently merged model for an example https://github.com/huggingface/transformers/pull/41112/files#diff-857421affc3c877bca95377cbb6adb3a8374b149fcbdcc6c759ea0408fa96897

zhang-prog committed:
Merge remote-tracking branch 'origin/main' into feat/paddleocr_vl

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1182 @@
+
+
+class PaddleOCRVLModel(Qwen2VLModel):
+    _checkpoint_conversion_mapping = {"^model": "language_model"}
```
The modification for  `_checkpoint_conversion_mapping` is here: https://github.com/huggingface/transformers/pull/42396

Now, we need to pass in `key_mapping` to ensure correct inference, which is a bit complicated. Is there a better way to do this?

example:

```diff
from transformers import AutoProcessor, AutoModelForImageTextToText

+ key_mapping = {
+    "^visual": "model.visual",
+    "^mlp_AR": "model.projector",
+    r"^model(?!(\.visual|\.projector))": "model.language_model",
+ }

processor = AutoProcessor.from_pretrained("PaddlePaddle/PaddleOCR-VL")
- model = AutoModelForImageTextToText.from_pretrained("PaddlePaddle/PaddleOCR-VL", dtype="bfloat16")
+ model = AutoModelForImageTextToText.from_pretrained("PaddlePaddle/PaddleOCR-VL", dtype="bfloat16", key_mapping=key_mapping)
messages = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/ocr_demo2.jpg"},
            {"type": "text", "text": "OCR:"},
        ]
    }
]
inputs = processor.apply_chat_template(
	messages,
	add_generation_prompt=True,
	tokenize=True,
	return_dict=True,
	return_tensors="pt",
).to(model.device)

outputs = model.generate(**inputs, max_new_tokens=100)
result = processor.decode(outputs[0][inputs["input_ids"].shape[-1]:-1])
print(result)
```

zhang-prog replied:
I understand:  
add `paddleocrvl` to `VLMS` in `transformers/src/transformers/conversion_mapping.py` to use `_checkpoint_conversion_mapping`

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1182 @@
+                grid_h,
+                self.patch_size,
+                grid_w,
+                self.patch_size,
```
Done

zhang-prog committed:
update

zhang-prog commented:
@zucchini-nlp 

PTAL.❤️

`_checkpoint_conversion_mapping` and `ignore_keys_at_rope_validation` needs to be discussed.

I am working on the docs and tests.

zucchini-nlp commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -273,26 +271,27 @@ def _preprocess(
-        assert self.temporal_patch_size == 1
-        flatten_patches = patches.reshape(grid_t * grid_h * grid_w, channel, self.patch_size, self.patch_size)
+        if temporal_patch_size != 1:
+            raise ValueError("temporal_patch_size must be 1!")
```
super nit: I'd frame it as `"temporal_patch_size must be 1, but got {temporal_patch_size}!"` to give more information

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1380 @@
+            image_embeds = self.get_image_features(pixel_values, image_grid_thw)
+            image_embeds = self.projector(image_embeds, image_grid_thw)
+            image_mask = self.get_placeholder_mask(input_ids, inputs_embeds=inputs_embeds, image_features=image_embeds)
+            inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
```
Done

zhang-prog committed:
fix unresolved problems

zhang-prog commented on src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py:
```diff
@@ -0,0 +1,1182 @@
+
+
+logger = logging.get_logger(__name__)
+rope_config_validation = partial(_rope_config_validation, ignore_keys={"mrope_section"})
```
ok. I will remove `kwargs["ignore_keys_at_rope_validation"] = {"mrope_section"}` firstly ; otherwise, the `make style` will not pass. 

I will adjust this part of the code according to the main branch (your subsequent modifications).

zhang-prog committed:
fix how position_ids work with flash_attn_2

zhang-prog committed:
Merge remote-tracking branch 'origin/main' into feat/paddleocr_vl

zhang-prog committed:
add tests and fix code

zhang-prog committed:
add model_doc

zhang-prog committed:
update model_doc

zhang-prog committed:
Merge remote-tracking branch 'origin/main' into feat/paddleocr_vl

zhang-prog committed:
fix ci

zhang-prog committed:
Merge remote-tracking branch 'origin/main' into feat/paddleocr_vl

zhang-prog committed:
update docstring

zhang-prog committed:
add tests

zhang-prog committed:
Merge remote-tracking branch 'origin/main' into feat/paddleocr_vl

zhang-prog committed:
update

zhang-prog committed:
add **kwargs

zhang-prog committed:
update

zhang-prog committed:
update

zucchini-nlp commented:
@zhang-prog worker crashed means that the tests might be using too much RAM. I see that the image sizes in tests are quite high, 600x400 images. Let's make dummy inputs and model as tiny as possible

Reviewing now

zucchini-nlp commented on docs/source/en/model_doc/paddleocr_vl.md:
```diff
@@ -0,0 +1,233 @@
+*This model was released on 2025.10.16 and added to Hugging Face Transformers on 2025.12.10*
```
nit: licence header here

zhang-prog replied:
Done

zucchini-nlp commented on utils/check_config_attributes.py:
```diff
@@ -58,6 +58,7 @@
         "expert_layer_offset",
         "expert_layer_period",
     ],
+    "PaddleOCRTextConfig": ["tie_word_embeddings"],
```
interesting, tie_word_embeddings is a universal attribute and I wouldn't expect CI to complain. Will check 👁️

zucchini-nlp replied:
Ah, I see. It has a comment that `# Allow if the default value in the configuration class is different from the one in PreTrainedConfig`, and there are more models skipping `tie_word_embeddings` explicitly

zucchini-nlp commented on tests/models/paddleocr_vl/test_modeling_paddleocr_vl.py:
```diff
@@ -0,0 +1,515 @@
+            "pad_token_id": 0,
+            "patch_size": 14,
+            "spatial_merge_size": 2,
+            # "torch_dtype": "bfloat16"
```
nit: let's delete commented part

zhang-prog replied:
Done

zucchini-nlp commented on tests/models/paddleocr_vl/test_modeling_paddleocr_vl.py:
```diff
@@ -0,0 +1,515 @@
+
+    @unittest.skip("PaddleOCRVL does not support this test.")
+    def test_num_layers_is_small(self):
+        pass
```
this test has to pass, it checks whether the model is small enough to not crash the runners 😄

zhang-prog committed:
update

zucchini-nlp commented:
run-slow: paddleocr_vl

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_42178). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp commented on docs/source/en/model_doc/paddleocr_vl.md:
```diff
@@ -0,0 +1,248 @@
+<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```
ultra nit: 2025

zucchini-nlp reviewed (approved):
Thanks a lot, huge work! Last bits will be adjusting the slow tests. Above comment shows failing cases

I will ping core maintainers in the meanwhile for review

zhang-prog committed:
reduce max_position_embeddings in tests

ArthurZucker commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -273,6 +273,7 @@
         ("ovis2", "Qwen2TokenizerFast" if is_tokenizers_available() else None),
         ("owlv2", "CLIPTokenizerFast" if is_tokenizers_available() else None),
         ("owlvit", "CLIPTokenizerFast" if is_tokenizers_available() else None),
+        ("paddleocr_vl", "LlamaTokenizer" if is_tokenizers_available() else None),
```
```suggestion
        ("paddleocr_vl", "TokenizersBackend" if is_tokenizers_available() else None),
```
from looking at the `tokenizer.json` on the hub, its not a llama!
```python
"normalizer": {
    "type": "Sequence",
    "normalizers": [
      {
        "type": "Replace",
        "pattern": {
          "String": " "
        },
        "content": "▁"
      }
    ]
  },
  "pre_tokenizer": null,
```

while llama would initialize: 
```python 
        self._tokenizer.normalizer = None
        self._tokenizer.pre_tokenizer = pre_tokenizers.Metaspace(
            replacement="▁", prepend_scheme=_get_prepend_scheme(self.add_prefix_space, self), split=False
        )
```

ArthurZucker reviewed (approved):
Very nice! 🤗

zucchini-nlp commented:
@zhang-prog the test changes hidden sizes of the model to be multiple of 16 iirc, prob that is the reason it is OOM'ing. If the model causes OOM and can't be made smaller, we can skip the test imo

zhang-prog committed:
update

zucchini-nlp commented:
run-slow: paddleocr_vl
